src/js/models/List.js

```diff
@@ -0,0 +1,204 @@
+   * @instance
+   */
+  destroy() {
+    for (var i = 0, n = this._items.length; i < n; i++) {
+      this._items[i].destroy()
+      delete this._items[i]
+    }
+    delete this._items
+  }
+
+  /**
+   * @returns {Record}  the record class itself
+   */
+  static getModel() {
+    return Model
+  };
+
+  /**
+   * @public
+   * @instance
+   * 
+   * @return {string} the list id
+   */
+  getID() {
+    return this.id
+  }
+
+  /**
+   * @public 
+   * @instance
+   * 
+   * @param {number} index the item index to retrieve 
+   * 
+   * @return {Model}       the Model object
+   */
+  get(index) {
+    return (this._items[index] || null)
+  }
+
+  /**
+   * @public
+   * @instance
+   * 
+   * @param {Model}     item    A list item to insert
+   * @param {number=-1} index   The index at which to insert the list, defaults to the end
+   */
+  insert(item, index = -1) {
+    if (!(item instanceof this.constructor.getModel())) {
+      throw new TypeError(`Expect an instance of ${this.constructor.getModel().name} in List.prototype.insert`)
+    }
+
+    index === -1
+
+      // insert a item at the end of no number specified
+      ? this.push(item)
+
+      // otherwise, insert at specified location
+      : this._items.splice(index, 0, item)
+  }
+
+  /**
+   * @public
+   * @instance
+   * 
+   * @param {Model} item An item to push on to the end of the items
+   */
+  push(item) {
+    if (!(item instanceof this.constructor.getModel())) {
+      throw new TypeError(`Expect an instance of ${this.constructor.getModel().name} in push`)
+    }
+
+    this._items.push(item)
+  }
+
+  /**
+   * @param {number} index   index of the item to remove
+   */
+  delete(index) {
+    this._items[index] && this._items.splice(index, 1)
+  }
+
+  /**
+   * Removes a range of items and returns them as a new List
+   * 
+   * @public
+   * @instance
+   * 
+   * @param {number} start    the start index
+   * @param {number} count    the number of items to slice off
+   * @param {List}   [target] the target list
+   * 
+   * @returns {List} A new list with the slice which is either the list or one created on the fly
+   */
+  sliceInto(start, count, target) {
+    let slice = this._items.splice(start, count)
+
+    // use or create the new List, and insert the slice into it
+    target = target || new this.constructor()
+    slice.forEach(item => target.insert(item))
+
+    return target
+  }
+
+  /**
+   * @public
+   * @instance
+   * 
+   * @return {number} the number of items
+   */
+  count() {
+    return this._items.length
+  }
+
+  /**
+   * @public
+   * @instance
+   * 
+   * @param {function} cb       A callback that receives each item and index
+   * @param {object}   context  The context within which to run the callback
+   */
+  forEach(cb, context = null) {
+    for (let i = 0, n = this._items.length; i < n; i++) {
+      context
+        ? cb(this._items[i], i, this._items)
+        : cb.call(context, this._items[i], i, this._items)
```

Comment:
I know nothing about how Javascript engines optimize, so perhaps this comment is ultimately moot.  The code itself suggests we will do a boolean check for every single item in the list (so n * this._items.length).  If we put the conditional outside the for loop  then we do the check only once (though we pay for creating a pointer):

const currentContext = context ? ... 
for (let i = 0 ....

Which solution is better depends on the number of elements in the array, the possibility that the Javascript engine already optimizes, and the cost of creating a pointer.  What are your thoughts?

Reply:
Actually, I just ran a whole bunch of looping perf stuff a week or so back, and this was  found most efficient.

This structure:
```javascript
for (let i = 0, n = array.length; i < n; i++) {
```
creates the `n` pointer at the same time as the `i` pointer, and does so only once. Whereas doing:
```javascript
for (let i = 0; i < array.length; i++) {
```
looks up a property on an array, which is slower than accessing a variable only.

Doing
```javascript
let n = array.length
for (let i = 0, ; i < n; i++) {
```
is functionally equivalent to the first, but the first is shorter and I don't think it compromises clarity (especially once you're used to seeing the pattern).

Essentially, the middle and final conditions are evaluated `n` times, but the first only once. So as little work as possible should be done in the latter two.

Reply:
Ah yes, now I see. I was just being dense.

Patching this up now. Good point.